ENH: Added template parameter for texture mask type

emmenlau opened the pull request:
Currently `ScalarImageToCooccurrenceMatrixFilter` assumes that the mask image would be the same type as the gray value image. This is not ideal because masks do not necessarily have to be of the same type. This limits the usability of the current design. What is more important is that forcing the mask to be of the same type can be problematic when working with floating-point types.

This PR extends `ScalarImageToCooccurrenceMatrixFilter` to add a template parameter for the mask image type, and uses the updated filter in `ScalarImageToTextureFeaturesFilter`.

This is my first contribution to ITK and I do not feel confident to say if the PR is breaking the API. It *does* add a new template parameter (with a default value) in front of an existing template parameter. So people that used both parameters before will not be able to build against this change.

Suggestions for how to change this PR would be very welcome!

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

dzenanz commented on Modules/Numerics/Statistics/include/itkScalarImageToCooccurrenceMatrixFilter.h:
```diff
@@ -90,7 +90,9 @@ namespace Statistics
 
-template <typename TImageType, typename THistogramFrequencyContainer = DenseFrequencyContainer2>
+template <typename TImageType,
+          typename TMaskType = TImageType,
```
I understand that this is a more logical sequence of template parameters. But for painless backwards compatibility it is way better to put the new parameter last. That way other people will not need to change their code.

dzenanz commented on Modules/Numerics/Statistics/include/itkScalarImageToTextureFeaturesFilter.h:
```diff
@@ -105,7 +105,9 @@ namespace Statistics
 
-template <typename TImageType, typename THistogramFrequencyContainer = DenseFrequencyContainer2>
+template <typename TImageType,
+          typename TMaskType = TImageType,
```
Change order, same as above.

blowekamp commented:
Looks like a very nice PR. Thank you for the contribution.

Given that we have `SetMaskImage` and `GetMaskImage` the expected typename is `MaskImageType`. Could you please update the `MaskType` to `MaskImageType`. Thank you.

emmenlau commented:
I've changed the `MaskType` to `MaskImageType` and `TMaskType` to `TMaskImageType`. Also, I've modified the test to actually use a different mask type than the image type, so test the new behavior. Everything builds and tests for me.

dzenanz commented:
There is now a [warning on Windows](https://open.cdash.org/viewBuildError.php?type=1&buildid=7564536): `itkScalarImageToCooccurrenceMatrixFilter.hxx(274): warning C4805: '!=': unsafe mix of type 'bool' and type 'unsigned char' in operation`. So your test expansion uncovered a problem with your patch 😄👍. `InsidePixelValue` should have the same type as mask image, not the intensity image. That should be an easy fix.

emmenlau commented:
> So your test expansion uncovered a problem with your patch smile+1. `InsidePixelValue` should have the same type as mask image, not the intensity image. That should be an easy fix.

Very well spotted! Fixed in latest version. While I was looking into this, I also saw that `region` is of type `ImageType::RegionType`  but used in the iterator for the image and the mask. Is that ok, or what is the clean way to deal with this? A static assert that ensures both image types have the same region type? Like
```
static_assert(std::is_same_v<typename ImageType::RegionType, typename MaskImageType::RegionType>, "error");
```

dzenanz commented:
If images have the same dimension, their region type will be the same. I think that same dimension is enforced by default somewhere in the processing pipeline.

dzenanz commented on Modules/Numerics/Statistics/include/itkScalarImageToCooccurrenceMatrixFilter.h:
```diff
@@ -215,7 +221,7 @@ class ITK_TEMPLATE_EXPORT ScalarImageToCooccurrenceMatrixFilter : public Process
   bool                  m_Normalize;
 
-  PixelType m_InsidePixelValue;
+  MaskPixelType m_InsidePixelValue;
```
On lines 184-185
```cpp
itkSetMacro(InsidePixelValue, PixelType);
itkGetConstMacro(InsidePixelValue, PixelType);
```
also needs to be changed to `MaskPixelType`.

dzenanz commented:
Do you plan to finish this PR @emmenlau?

emmenlau commented:
Could you please take a look again? Thanks for the review!
